rfcs/2021-09-01-8547-accept-metrics-in-datadog-agent-source.md

```diff
@@ -0,0 +1,168 @@
+  the beginning it will also proxify `/api/v1/series` & `/api/beta/sketches`.
+- Handle the `/api/v1/series` route (based on both the offical [API](https://docs.datadoghq.com/api/latest/metrics/) and
+  the [Datadog Agent itself](https://github.com/DataDog/datadog-agent/blob/main/pkg/forwarder/telemetry.go#L20-L31)) to
+  cover every metric type handled by this endpoint (count, gauge and rate) and:
+    - Add support for missing fields in the `datadog_metrics` sinks
+    - Align tag manipulation (Datadog allows `key:foo` & `key:bar`, and the tags are encoded as a list whereas Vector
+      encode those as a map).
+    - Overall this is fairly straighforward except for the tagging issue
+- Handle the `/api/beta/sketches` route in the `datadog_agent` source to support sketches/distribution encoded using
+  protobuf
+   - Revamp the distribution metrics in the `datadog_metrics` sink to use sketches and the associated endpoint will
+     probably be a good idea.
+   - The sketches the agent ships is based on this [paper](http://www.vldb.org/pvldb/vol12/p2195-masson.pdf) whereas
+     Vector uses what's called a summary inside the Agent, implementing the complete DDSketch support in Vector is
+     probably a good idea as sketches have convenient properties for wide consistent aggregation and limited error. To
+     support smooth migration, full DDsktech support is mandatory, as customers that emit distribution metric from
+     Datadog Agent would need it to migrate to Vector aggegation.
+
+**Regarding the tagging issue:** A -possibly temporary- work-around would be to store incoming tags with the complete
+"key:value" string as the key and an empty value to store those in the extisting map Vector uses to store
+[tags](https://github.com/timberio/vector/blob/master/lib/vector-core/src/event/metric.rs#L60) and slightly rework the
+`datadog_metrics` sink not to append `:` if a tag key has the empty string as the corresponding value.
+
+
+## Rationale
+
+- Smoother Vector integration with Datadog.
+- Needed for Vector to act as a complete Datadog Agent aggregator (but further work will still be required).
+- Extend the Vector ecosystem, bring additional feature for distribution metrics that would enable consistent
+  aggregation.
+
+## Drawbacks
+
+- It arbitrary imposes internal sketches representation to follow the DDSketch paper.
+
+## Prior Art
+
+- Existing Vector metrics source over HTTP(s) works well, those constructs will be reused
+- Datadog logs is already supported in the `datadog_agent` source
+- Vector has its own vector-to-vector protocol that serve a similar purpose but the idea is to enable Vector to receive
+  Datadog metrics from unmodified Datadog Agents so that is probably not an option
+- There are many [ways](https://edoliberty.github.io/papers/streamingQuantiles.pdf) of implementing quantile sketches
+
+## Alternatives
```

Comment:
The override settings are not consistent, especially as those we want to deal with here are the one with the longest history. I think Agent past versions along with some backend history (and future) explain this situation. I don't think there would be any blocker to add more selective override however the one thing that would worry me in that situation is that some metric route may evolve over time (e.g. `/api/beta/sketches` is the very first piece of v2 API that might appear at some point), and we would need to get all Vector user upgrade to support that new route before they upgrade this agent. A simple solution would be to proxify unknow route to Vector but that bring us to the solution I suggest. Another solution -a bit hacky IMHO- would be to leverage the haproxy setup documented [here](https://docs.datadoghq.com/agent/proxy/?tab=agentv6v7#haproxy) and selectively divert route to vector, while it's not what we could call a clean & simple deployment it would pretty much solve all those kind of problem.
I think the discussion about which entity implements the route flexibility we need is interesting, either we do it in Vector with a shim proxy layer, or in the Datadog Agent with new `xxx_dd_url` overrides or we outsource that to a middle. I added those consideration into the RFCs, but IMHO we should always expect that legitimate-but-unsupported-by-vector metric payload will be received by Vector at some point for various reasons.